routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,285 @@
+)
+
+const artifactRouteBase = "/_apis/pipelines/workflows/{taskID}/artifacts"
+
+func ArtifactsRoutes(goctx gocontext.Context, prefix string) *web.Route {
+	m := web.NewRoute()
+	m.Use(withContexter(goctx))
+
+	r := artifactRoutes{
+		prefix: prefix,
+		fs:     storage.ActionsArtifacts}
+
+	// retrieve, list and confirm artifacts
+	m.Post(artifactRouteBase, r.getUploadArtifactURL)
+	m.Get(artifactRouteBase, r.listArtifacts)
+	m.Patch(artifactRouteBase, r.patchArtifact)
+
+	// handle container artifacts
+	m.Put(artifactRouteBase+"/{artifactID}/upload", r.uploadArtifact)
+	m.Get(artifactRouteBase+"/{artifactID}/path", r.getDownloadArtifactURL)
+	m.Get(artifactRouteBase+"/{artifactID}/download", r.downloadArtifact)
+
+	return m
+}
+
+// withContexter initializes a package context for a request.
+func withContexter(ctx gocontext.Context) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+			ctx := context.Context{
+				Resp: context.NewResponse(resp),
+				Data: map[string]interface{}{},
+			}
+			defer ctx.Close()
+
+			ctx.Req = context.WithContext(req, &ctx)
+
+			next.ServeHTTP(ctx.Resp, ctx.Req)
+		})
+	}
+}
+
+type artifactRoutes struct {
+	prefix string
+	fs     storage.ObjectStorage
+}
+
+func (ar artifactRoutes) openFile(fpath string, contentRange string) (storage.Object, bool, error) {
+	if contentRange != "" && !strings.HasPrefix(contentRange, "bytes 0-") {
+		f, err := ar.fs.Open(fpath)
+		if err != nil {
+			return nil, false, err
+		}
+		_, err = f.Seek(0, os.SEEK_END)
+		return f, true, err
+	}
+	f, err := ar.fs.Open(fpath)
+	return f, false, err
+}
+
+func (ar artifactRoutes) buildArtifactUrl(taskID int64, artifactID int64, suffix string) (string, error) {
+	uploadURL := strings.TrimSuffix(setting.AppURL, "/") + strings.TrimSuffix(ar.prefix, "/") +
+		strings.ReplaceAll(artifactRouteBase, "{taskID}", strconv.FormatInt(taskID, 10)) +
+		"/" + strconv.FormatInt(artifactID, 10) + "/" + suffix
+	u, err := url.Parse(uploadURL)
+	if err != nil {
+		return "", err
+	}
+	return u.String(), nil
+}
+
+// getUploadArtifactURL generates a URL for uploading an artifact
+func (ar artifactRoutes) getUploadArtifactURL(ctx *context.Context) {
+	// get task
+	taskID := ctx.ParamsInt64("taskID")
+	task, err := actions.GetTaskByID(ctx, taskID)
+	if err != nil {
+		log.Error("Error getting task: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	artifact, err := actions.CreateArtifact(ctx, task)
+	if err != nil {
+		log.Error("Error creating artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+	url, err := ar.buildArtifactUrl(taskID, artifact.ID, "upload")
+	if err != nil {
+		log.Error("Error parsing upload URL: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	ctx.JSON(200, map[string]interface{}{
```

Comment:
A nit, all HTTP status codes has been refactored to `http.Xxx` const, so I think it's better to use constants from then on for consistency.

Reply:
All status code are instead of http.XXX constants